core/app/models/spree/taxon.rb

```diff
@@ -85,6 +85,16 @@ def active_products
       products.not_deleted.available
     end
 
+    # @return [ActiveRecord::Relation<Spree::Product>] all self and descendant products
+    def all_products
+      Product.joins(:taxons).where(spree_taxons: { id: descendant_ids + [id] })
```

Comment:
I have some query optimizations I think would be suitable here. Any time we use pluck, we take back the identifiers and put them in a ruby array which means a round trip to the database.

Currently `all_products` is required to do 2 queries, and `all_variants` is required to do 3, but we can reduce that to one each pretty easily.

Example of current behaviour:

```
[1] pry(main)> Spree::Taxon.first.all_variants.count
  Spree::Taxon Load (0.1ms)  SELECT  "spree_taxons".* FROM "spree_taxons" ORDER BY "spree_taxons"."id" ASC LIMIT ?  [["LIMIT", 1]]
   (0.2ms)  SELECT "spree_taxons"."id" FROM "spree_taxons" WHERE ("spree_taxons"."lft" >= 1) AND ("spree_taxons"."lft" < 12) AND ("spree_taxons"."id" != 1) ORDER BY "spree_taxons"."lft"
   (0.2ms)  SELECT "spree_products"."id" FROM "spree_products" INNER JOIN "spree_products_taxons" ON "spree_products_taxons"."product_id" = "spree_products"."id" INNER JOIN "spree_taxons" ON "spree_taxons"."id" = "spree_products_taxons"."taxon_id" WHERE "spree_products"."deleted_at" IS NULL AND "spree_taxons"."id" IN (3, 4, 5, 6, 7, 1)
   (0.1ms)  SELECT COUNT(*) FROM "spree_variants" WHERE "spree_variants"."deleted_at" IS NULL AND "spree_variants"."product_id" IN (1, 2, 8, 9, 4, 3, 5, 7, 6)
```

But if we try this:
```Ruby
    # @return [ActiveRecord::Relation<Spree::Product>] all self and descendant
    #   products
    def all_products
      scope = Product.joins(:taxons)
      scope.where(
        spree_taxons: { id: descendants.select(:id) }
      ).or(scope.where(spree_taxons: { id: id } ))
    end

    # @return [ActiveRecord::Relation<Spree::Variant>] all self and descendant
    #   variants, including master variants.
    def all_variants
      Variant.where(product_id: all_products.select(:id))
    end
```

Everything can be done in a single subquery:

```[3] pry(main)> Spree::Taxon.first.all_variants.count
  Spree::Taxon Load (0.2ms)  SELECT  "spree_taxons".* FROM "spree_taxons" ORDER BY "spree_taxons"."id" ASC LIMIT ?  [["LIMIT", 1]]
   (0.3ms)  SELECT COUNT(*) FROM "spree_variants" WHERE "spree_variants"."deleted_at" IS NULL AND "spree_variants"."product_id" IN (SELECT "spree_products"."id" FROM "spree_products" INNER JOIN "spree_products_taxons" ON "spree_products_taxons"."product_id" = "spree_products"."id" INNER JOIN "spree_taxons" ON "spree_taxons"."id" = "spree_products_taxons"."taxon_id" WHERE ("spree_products"."deleted_at" IS NULL AND "spree_taxons"."id" IN (SELECT "spree_taxons"."id" FROM "spree_taxons" WHERE ("spree_taxons"."lft" >= 1) AND ("spree_taxons"."lft" < 12) AND ("spree_taxons"."id" != 1) ORDER BY "spree_taxons"."lft") OR "spree_products"."deleted_at" IS NULL AND "spree_taxons"."id" = ?))  [["id", 1]]
=> 19
```

This means only one round trip to the DB, less work done in Ruby and hopefully happy developers!